feat: add tag display on item card

djaustin opened the pull request.

djaustin committed:
feat: add tag display on item card

battmeech commented on src/campaign/components/ItemCard/TagGroup.tsx:
```diff
@@ -0,0 +1,55 @@
+  colorScheme,
+  ...wrapProps
+}) => {
+  const [showAll, setShowAll] = useBoolean(false);
```
How's this different from useState(false)?

djaustin replied:
It just gives some more useful functions specifically tailored to booleans. These are in the setShowAll obect.

Below I'm using `onClick={setShowAll.toggle}` instead of `onClick={() => setShowAll(!showAll)}` it also provides one-way functions like setShowAll.off and setShowAll.on

battmeech reviewed (commented):
Looks good to me so far!

Let me know when you're done and I'll take another look 😁

djaustin commented:
![2021-08-14 15-31-32 2021-08-14 15_35_06](https://user-images.githubusercontent.com/9084497/129449910-b12cd057-22a3-40e1-a390-c2b33b706dc8.gif)

djaustin committed:
feat: add tags to edit and add item modal

'downlevelIteration' tsconfig allows Set objects
to be spread when targetting ES5

battmeech commented on src/campaign/components/ItemCard/__test__/TagGroup.test.tsx:
```diff
@@ -0,0 +1,50 @@
+  describe('"show all tags" button', () => {
+    it("should be visible when limit is less than tag count", () => {
+      const limit = 3;
+      render(<TagGroup tags={tags} displayLimit={limit} />);
```
Can we split up tests into 3 blocks, makes them easier to read. 

Setup

Run

Assert

battmeech commented on src/campaign/components/ItemModal/ItemForm.tsx:
```diff
@@ -10,6 +10,8 @@ import {
   VStack,
 } from "@chakra-ui/react";
+import React from "react";
+import TagInput from "./TagInput";
```
Can we avoid default exports for consistency?

battmeech commented on src/campaign/components/ItemModal/TagInput.tsx:
```diff
@@ -0,0 +1,61 @@
+
+export type TagInputProps = {
+  // eslint-disable-next-line no-unused-vars
+  onTagsChanged?: (tags: string[]) => void;
```
General comment - there must be a way to configure the linter to understand this isn't a problem. I've had this issue a few times.

djaustin committed:
feat: include tag content in filter

battmeech commented on src/campaign/components/ItemModal/__test__/ItemForm.test.tsx:
```diff
@@ -1,3 +1,4 @@
+import userEvent from "@testing-library/user-event";
```
How long do these tests take to run using userEvent vs fireEvent?

djaustin replied:
The tests in this file take ~100ms at the slowest. userEvent is super useful for typing text. I don't even know how you'd achieve that with fireEvent. Probably a series of `.keyPress` calls for each character.

battmeech commented on src/campaign/components/ItemModal/useItemForm.tsx:
```diff
@@ -14,19 +14,23 @@ type Item = {
-  setValues: (value: { key: keyof Item; value: string | number }) => void;
+  setValues: (value: {
+    key: keyof Item;
+    value: string | number | string[];
```
We might need to add a ticket to try and refactor this hook, starting to become quite complex with these multiple types!

djaustin replied:
Agreed!

battmeech commented on tsconfig.json:
```diff
@@ -17,7 +17,8 @@
     "jsx": "preserve",
-    "baseUrl": "src"
+    "baseUrl": "src",
+    "downlevelIteration": true
```
What's this do?

djaustin replied:
Allows spread operator on `Set` objects. I think we need this because we're targeting ES5 not ES6 so it adds some compatibility for older javascript versions. I've added a brief description in the commit message too

battmeech reviewed (approved):
Looks good. A few small nitpicks sorry!

djaustin committed:
refactor: explicitly structure tests and remove default exports
